File: 프로그래머스/lv2/17677. ［1차］ 뉴스 클러스터링/［1차］ 뉴스 클러스터링.py
def solution(str1, str2):
    str1=str1.lower()
    str2=str2.lower()
    possible_str = set()
    str1_dict = {}
    str2_dict = {}
    intersection=0
    union=0
    for i in range(len(str1)-1):
        if not str1[i:i+2].isalpha():
            continue
        possible_str.add(str1[i:i+2])
        if str1[i:i+2] not in str1_dict:
            str1_dict[str1[i:i+2]]=0
        str1_dict[str1[i:i+2]]+=1
        
    for i in range((len(str2)-1)):
        if not str2[i:i+2].isalpha():
            continue
        possible_str.add(str2[i:i+2])
        if str2[i:i+2] not in str2_dict:
            str2_dict[str2[i:i+2]]=0
        str2_dict[str2[i:i+2]]+=1
    
    possible_str = list(possible_str)
    
    for s in possible_str:
        if s in str1_dict and s in str2_dict:
            if str1_dict[s]<=str2_dict[s]: 
                union+=str2_dict[s]
                intersection+=str1_dict[s]
            else:
                union+=str1_dict[s]
                intersection+=str2_dict[s]
        elif s in str1_dict:
            union+=str1_dict[s]
        else:
            union+=str2_dict[s]
    
    if union==0:
        answer=65536
    else:
        answer = (intersection/union)*65536//1
    return answer

File: 프로그래머스/2/17677. ［1차］ 뉴스 클러스터링/［1차］ 뉴스 클러스터링.java
import java.util.*;

class Solution {
    public int solution(String str1, String str2) {
        
        
        List<String> list1 = generateList(str1);
        List<String> list2 = generateList(str2);
        
        
        int intersection = countIntersection(list1, list2);
        int union = countUnion(list1, list2);
        
        if(intersection==0 && union==0){
            return 65536;
        }
        
        int answer = (int)(intersection*65536/(double)union);
        return answer;
    }
    
    private int countUnion(List<String> list1, List<String> list2){
        int count=0;
        boolean[] vis = new boolean[list2.size()];
        for(int i=0; i<list1.size(); i++){
            for(int j=0; j<list2.size(); j++){
                if(vis[j]) continue;
                if(list1.get(i).equals(list2.get(j))){
                    vis[j]=true;
                    break;
                }
            }
        }
        
        count = list1.size();
        for(int i=0; i<list2.size(); i++){
            if(vis[i]) continue;
            count++;
        }
        
        return count;
    }
    
    private int countIntersection(List<String> list1, List<String> list2){
        int count=0;
        boolean[] vis = new boolean[list2.size()];
        for(int i=0; i<list1.size(); i++){
            for(int j=0; j<list2.size(); j++){
                if(vis[j]) continue;
                if(list1.get(i).equals(list2.get(j))){
                    count++;
                    vis[j]=true;
                    break;
                }
            }
        }
        return count;
    }
    
    
    private List<String> generateList(String str){
        List<String> list = new ArrayList<>();
        for(int i=1; i<str.length(); i++){
            if(isAlphabet(str.charAt(i-1)) && isAlphabet(str.charAt(i))){
                list.add(str.substring(i-1,i+1).toLowerCase());
            }
        }
        return list;
    }
    
    private boolean isAlphabet(char c){
        return (c>='a' && c<='z') || (c>='A' && c<='Z');
    }
}